File: benchmarks/python/echo_client.py
# ---------------------------------------------------------------------------
# PROPRIETARY CODE – Arthur de Araújo Farias 2025
# All rights reserved.  No part of this file may be reproduced, stored in a
# retrieval system, or transmitted in any form or by any means—electronic,
# mechanical, photocopying, recording, or otherwise—without the prior written
# permission of the copyright holder.
# ---------------------------------------------------------------------------

"""asyncio closed-loop load generator for the cross-stack benchmark. One
event loop; --connections sockets each keeping --pipeline frames in flight.
Emits one JSON result line (see common.py).

    python echo_client.py --proto tcp --host 127.0.0.1 --port 7001 \
        --connections 64 --duration 10 --warmup 3 --payload 64 --pipeline 1
"""

from __future__ import annotations

import asyncio
import sys
import time

from common import Histogram, emit_result, parse_args

o = parse_args(sys.argv[1:])
hist = Histogram()
st = {"measuring": False, "running": True, "requests": 0, "errors": 0}
now_ns = time.monotonic_ns


def make_frame() -> bytes:
    return now_ns().to_bytes(8, "little") + b"\0" * (o.payload - 8)


LATE_NS = 250_000_000  # a straggler (dropped datagram), not a latency sample


def book(ts_ns: int) -> None:
    rtt = now_ns() - ts_ns
    if rtt > LATE_NS:
        st["errors"] += 1
        return
    if st["measuring"]:
        hist.record(rtt)
        st["requests"] += 1


class TcpClient(asyncio.Protocol):
    def connection_made(self, transport: asyncio.Transport) -> None:
        try:
            transport.get_extra_info("socket").setsockopt(6, 1, 1)  # TCP_NODELAY
        except OSError:
            pass
        self.transport = transport
        self.buf = bytearray()
        self.inflight = 0
        self.pump()

    def pump(self) -> None:
        while st["running"] and self.inflight < o.pipeline:
            self.transport.write(make_frame())
            self.inflight += 1

    def data_received(self, data: bytes) -> None:
        self.buf += data
        p = o.payload
        while len(self.buf) >= p:
            ts = int.from_bytes(self.buf[:8], "little")
            del self.buf[:p]
            self.inflight -= 1
            book(ts)
            self.pump()

    def connection_lost(self, exc) -> None:
        if st["running"]:
            st["errors"] += 1


class UdpClient(asyncio.DatagramProtocol):
    def connection_made(self, transport: asyncio.DatagramTransport) -> None:
        self.transport = transport
        self.inflight = 0
        self.last_rx = now_ns()
        self.pump()

    def pump(self) -> None:
        while st["running"] and self.inflight < o.pipeline:
            self.transport.sendto(make_frame())
            self.inflight += 1
        self.last_rx = now_ns()

    def datagram_received(self, data: bytes, addr) -> None:
        if len(data) >= 8:
            book(int.from_bytes(data[:8], "little"))
        if self.inflight > 0:
            self.inflight -= 1
        self.last_rx = now_ns()
        self.pump()

    def error_received(self, exc) -> None:
        st["errors"] += 1


async def main() -> None:
    loop = asyncio.get_running_loop()
    protos: list = []

    if o.proto == "udp":
        for _ in range(o.connections):
            _, proto = await loop.create_datagram_endpoint(UdpClient, remote_addr=(o.host, o.port))
            protos.append(proto)

        async def sweeper() -> None:
            while st["running"]:
                await asyncio.sleep(0.1)
                now = now_ns()
                for pr in protos:
                    if pr.inflight > 0 and now - pr.last_rx > 200_000_000:
                        st["errors"] += pr.inflight
                        pr.inflight = 0
                    pr.pump()

        sweep_task = asyncio.create_task(sweeper())
    else:
        for _ in range(o.connections):
            _, proto = await loop.create_connection(TcpClient, o.host, o.port)
            protos.append(proto)
        sweep_task = None

    await asyncio.sleep(o.warmup)
    st["measuring"] = True
    t0 = now_ns()
    await asyncio.sleep(o.duration)
    st["measuring"] = False
    st["running"] = False
    elapsed = (now_ns() - t0) / 1e9
    await asyncio.sleep(0.1)

    if sweep_task:
        sweep_task.cancel()
    for pr in protos:
        pr.transport.close()

    emit_result(o, hist, st["requests"], st["errors"], elapsed)


if __name__ == "__main__":
    asyncio.run(main())

File: benchmarks/node/echo-client.js
// ---------------------------------------------------------------------------
// PROPRIETARY CODE – Arthur de Araújo Farias 2025
// All rights reserved.  No part of this file may be reproduced, stored in a
// retrieval system, or transmitted in any form or by any means—electronic,
// mechanical, photocopying, recording, or otherwise—without the prior written
// permission of the copyright holder.
// ---------------------------------------------------------------------------

// Node.js closed-loop load generator for the cross-stack benchmark. One
// event loop; `--connections` sockets each keeping `--pipeline` frames in
// flight. Emits one JSON result line (see ./common.js).
//
//   node echo-client.js --proto tcp --host 127.0.0.1 --port 7001 \
//       --connections 64 --duration 10 --warmup 3 --payload 64 --pipeline 1

'use strict';
const net = require('node:net');
const dgram = require('node:dgram');
const { parseArgs, Histogram, emitResult } = require('./common');

const o = parseArgs(process.argv.slice(2));
const hist = new Histogram();
const state = { measuring: false, running: true, requests: 0, errors: 0 };
const nowNs = () => process.hrtime.bigint();

function makeFrame() {
  const b = Buffer.alloc(o.payload);
  b.writeBigUInt64LE(process.hrtime.bigint(), 0);
  return b;
}

const LATE_NS = 250_000_000n; // a straggler (dropped datagram), not a latency sample

function book(tsNs) {
  const rtt = process.hrtime.bigint() - tsNs;
  if (rtt > LATE_NS) { state.errors++; return; }
  if (state.measuring) {
    hist.record(Number(rtt));
    state.requests++;
  }
}

function startTcp(done) {
  const conns = [];
  for (let i = 0; i < o.connections; i++) {
    const sock = net.connect({ host: o.host, port: o.port, noDelay: true });
    let acc = Buffer.alloc(0);
    let inflight = 0;
    const pump = () => {
      while (state.running && inflight < o.pipeline) { sock.write(makeFrame()); inflight++; }
    };
    sock.on('connect', pump);
    sock.on('error', () => { state.errors++; });
    sock.on('data', (chunk) => {
      acc = acc.length ? Buffer.concat([acc, chunk]) : chunk;
      while (acc.length >= o.payload) {
        const ts = acc.readBigUInt64LE(0);
        acc = acc.subarray(o.payload);
        inflight--;
        book(ts);
        pump();
      }
    });
    conns.push(sock);
  }
  done(() => { for (const s of conns) s.destroy(); });
}

function startUdp(done) {
  const flows = [];
  for (let i = 0; i < o.connections; i++) {
    const sock = dgram.createSocket('udp4');
    const f = { sock, inflight: 0, lastRx: 0n };
    const pump = () => {
      while (state.running && f.inflight < o.pipeline) { sock.send(makeFrame(), o.port, o.host); f.inflight++; }
      f.lastRx = process.hrtime.bigint();
    };
    sock.on('message', (msg) => {
      if (msg.length >= 8) book(msg.readBigUInt64LE(0));
      if (f.inflight > 0) f.inflight--;
      f.lastRx = process.hrtime.bigint();
      pump();
    });
    sock.on('error', () => { state.errors++; });
    f.pump = pump;
    flows.push(f);
  }
  const sweep = setInterval(() => {
    const now = process.hrtime.bigint();
    for (const f of flows) {
      if (f.inflight > 0 && f.lastRx && now - f.lastRx > 200_000_000n) { state.errors += f.inflight; f.inflight = 0; }
      f.pump();
    }
  }, 100);
  for (const f of flows) f.pump();
  done(() => { clearInterval(sweep); for (const f of flows) f.sock.close(); });
}

const start = o.proto === 'udp' ? startUdp : startTcp;
start((teardown) => {
  setTimeout(() => {
    state.measuring = true;
    const t0 = nowNs();
    setTimeout(() => {
      state.measuring = false;
      state.running = false;
      const elapsed = Number(nowNs() - t0) / 1e9;
      setTimeout(() => {
        teardown();
        emitResult(o, hist, state.requests, state.errors, elapsed);
        process.exit(0);
      }, 100);
    }, o.duration * 1000);
  }, o.warmup * 1000);
});
